API / Cache: Improve schema memory resources

joselcvarela opened the pull request:
## Scope
When doing some tests with a huge schema (>6000 entities, collections + fields + relations) I have found out that my Directus instance was running out of memory.
<details><summary>
To better understand what was causing this issue I have used Memory inspector and took some snapshots.
</summary>
I am not able to include them here as it has private data
</details>

When I took a look into the memory snapshot, I saw the schema repeated multiple times but without some properties.
I have found out that the [setLocalSchemaCache](https://github.com/joselcvarela/directus/blob/api/cache/improve-schema-memory/api/src/cache.ts#L121-L125) was the culprit.
More specifically, the schema is being cached in memory using Keyv.
But Keyv does serialize / deserialize all data that passes through it.
Since there's no specification, the default serializer / deserializer are being used.
[As we can see on the serializer](https://github.com/jaredwray/keyv/blob/main/packages/serialize/src/index.ts#L5-L53) it is traversing the entire object and store chunks of it as strings in memory.
This causes the data being duplicated in memory.
But this is not even necessary, because the schema is already in memory, so we just need to store it's reference, no need to serialize it.

In this PR, we just change the serialize / deserialize to just store the reference of the schema object.
Since Keyv is using a [Map](https://github.com/jaredwray/keyv/blob/main/packages/keyv/src/index.ts#L106) there should be no issues.

What's changed:

- Improvement on memory consumption storing schema into memory

## Potential Risks / Drawbacks

- If Keyv decides to use anything else other than a `Map` or any other structure that can hold object, this may break

## Review Notes / Questions

- Testing this can be somehow difficult but I am leaving some steps below

## Testing
In order to test this fix, I have:
1. Run Postgres and Redis via Docker Compose
2. Run bootstrap
3. Use cache, cache system and auto purge cache
4. Use Directus 10.13.3 via Docker
5. Use command `node --inspect=0.0.0.0 cli.js start`. This allows to investigate Memory.
6. Use DevTools to track memory usage
7. Run K6 to fetch data from multiple collections
8. At same time, do multiple sort fields in Data Model
9. See memory increasing and container run out of memory
10. Build a new image from 7dc377f8fd (tag v10.13.3) but with the fix implemented
11. Use new build image and repeat process from 5 until 8
12. After doing multiple sort fields (more than previous) the container was still running as expected with low memory
___

Fixes #24035
Closes SER-440

joselcvarela committed:
Do not serialize/deserialize if cache is memory
Since the object is in memory already, we do not need to have to serialize/deserialize.
Instead we just save the reference of the object avoiding memory and processing resources

joselcvarela committed:
Create dry-ants-explode.md

licitdev committed:
Add test

licitdev committed:
Update wording

licitdev commented on .changeset/dry-ants-explode.md:
```diff
@@ -0,0 +1,5 @@
+"@directus/api": patch
+---
+
+Improved memory and performance on saving schema locally
```
```suggestion
Optimized serialization and deserialization for memory cache
```

joselcvarela committed:
Don't cast to unexpected types

joselcvarela committed:
Retrieve a clone of schema
Because there's some places where schema is mutated.
Since schema is now shared between requests, we need to retrieve a clone in order to not mutate schema.

joselcvarela committed:
Use cloneDeep instead
structuredClone does not handle Buffers

joselcvarela committed:
Update .changeset/dry-ants-explode.md

Co-authored-by: ian <licitdev@gmail.com>

licitdev committed:
Test if cached value could be updated

licitdev commented on api/src/cache.ts:
```diff
@@ -163,6 +164,13 @@ function getConfig(store: Store = 'memory', ttl: number | undefined, namespaceSu
 
+	if (store === 'memory') {
+		config.compression = {
+			serialize: (v) => v,
+			deserialize: (v) => cloneDeep(v),
```
```suggestion
			serialize: (v: any) => cloneDeep(v),
			deserialize: (v: any) => cloneDeep(v),
```

licitdev reviewed (changes requested):
@joselcvarela Would you mind testing to confirm if the updated serialization and deserialization are more optimized than the original? 🤔

licitdev committed:
Merge branch 'main' into api/cache/improve-schema-memory

licitdev committed:
Scope optimisation only for local schema cache

ComfortablyCoding reviewed (commented):
Initial review regarding object.freeze usage

ComfortablyCoding commented on api/src/cache.ts:
```diff
@@ -65,7 +65,11 @@ export function getCache(): {
 	if (localSchemaCache === null) {
-		localSchemaCache = getKeyvInstance('memory', getMilliseconds(env['CACHE_SYSTEM_TTL']), '_schema');
+		localSchemaCache = getKeyvInstance('memory', getMilliseconds(env['CACHE_SYSTEM_TTL']), '_schema', {
+			serialize: (v: any) => Object.freeze(v),
```
Although we do not seem to directly modify the schema we do expose it to extensions. As such this is potentially a breaking change as freezing the object will cause unexpected results for anyone who is modifying it.

Additionally this change requires all contributors to be aware that the schema is frozen when working in the codebase

licitdev committed:
Merge branch 'main' into api/cache/improve-schema-memory

licitdev committed:
Try freezing parts of schema overview to allow "unfreezing"

licitdev commented:
Closed in favor of #24636
